### PR TITLE
HTTP-119 - Fix bug where default Java TrustStore was not used when no cert-related properties were used.

### PR DESCRIPTION
**Pull request by kristoffSC**
#### Description

Fix bug where default Java TrustStore was not used when no cert-related properties were used.

Resolves [Public certificates should not have to be supplied as they should be picked up from the jvm](https://github.com/getindata/flink-http-connector/issues/119)

##### PR Checklist
- [x] Tests added

### CODE REVIEW

**Comment by kristoffSC** [comment]
@grzegorz8 @davidradl  could you take a look?

@davidradl I Think this should solve the problem you reported.
Could you check it from the branch?

File: src/test/java/com/getindata/connectors/http/internal/table/lookup/querycreators/ElasticSearchLiteQueryCreatorTest.java [hunk 1]
```diff
@@ -18,6 +18,24 @@
 
 public class ElasticSearchLiteQueryCreatorTest {
 
+    @Test
```
Comment (grzegorz8) [inline src/test/java/com/getindata/connectors/http/internal/table/lookup/querycreators/ElasticSearchLiteQueryCreatorTest.java:21]: I think these new tests are not related to the issue HTTP-119? Could you please add them in a separate PR?

Comment (davidradl) [reply src/test/java/com/getindata/connectors/http/internal/table/lookup/querycreators/ElasticSearchLiteQueryCreatorTest.java:21]: Agreed - it would be good to have unit tests to test the various code paths around this change.

Comment (kristoffSC) [reply src/test/java/com/getindata/connectors/http/internal/table/lookup/querycreators/ElasticSearchLiteQueryCreatorTest.java:21]: The "new code" is already partially covered by existing tests, but I will try to add new ones to cover it fully.

The reason why I added those new tests was that we dropped with branch coverage below 90% and while inspecting test report I have found those two.

I will like to leave those tests in this PR, unless there will be a strong NO to it. The PR is small so IMO having those extra, unrelated tests is ok.

image

Comment (grzegorz8) [reply src/test/java/com/getindata/connectors/http/internal/table/lookup/querycreators/ElasticSearchLiteQueryCreatorTest.java:21]: That's fine for me.

**Review by grzegorz8 — APPROVED** [review]

File: src/main/java/com/getindata/connectors/http/internal/utils/JavaNetHttpClientFactory.java [hunk 2]
```diff
@@ -88,6 +91,21 @@ private static SSLContext getSslContext(Properties properties) {
         String clientPrivateKey = properties
             .getProperty(HttpConnectorConfigConstants.CLIENT_PRIVATE_KEY, "");
 
+        if (StringUtils.isNullOrWhitespaceOnly(keyStorePath)
+            && !selfSignedCert
+            && serverTrustedCerts.length == 0
+            && StringUtils.isNullOrWhitespaceOnly(clientCert)
+            && StringUtils.isNullOrWhitespaceOnly(clientPrivateKey)) {
```
Comment (davidradl) [inline src/main/java/com/getindata/connectors/http/internal/utils/JavaNetHttpClientFactory.java:98]: What happens if we only have client certs. In this case, we want the trusted public server certs and the client certs. The current fix does not deal with that combination. Can we get hold of the default public server certs and add them like  `securityContext.addCertToTrustStore(cert);  ` then process the client certs?

**Review by davidradl — COMMENTED** [review]
please see the comment I left in the code.

**Review by davidradl — CHANGES_REQUESTED** [review]
Could you have a look a the comment I left in the code please

File: src/main/java/com/getindata/connectors/http/internal/utils/JavaNetHttpClientFactory.java [hunk 3]
```diff
@@ -88,6 +91,21 @@ private static SSLContext getSslContext(Properties properties) {
...
+
+            try {
+                return SSLContext.getDefault();
+            } catch (NoSuchAlgorithmException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        SecurityContext securityContext = createSecurityContext(properties);
```
Comment (davidradl) [inline src/main/java/com/getindata/connectors/http/internal/utils/JavaNetHttpClientFactory.java:107]: Talking with our team around this, in addition to the above. It would be good to have the ability to:
1- run with public certs and any supplied private server certs. i.e. merge the supplied certs with existing public ones.
2- run only with private server certs (which is the current functionality when they are supplied in the config) 

This would require a new config option maybe 'gid.connector.http.security.cert.server.includeDefaultCerts' with a default of true. 
 
WDYT?

**Comment by davidradl** [comment]
@kristoffSC Friendly nudge. I assume this one is going to be processed before https://github.com/getindata/flink-http-connector/pull/117. We are wanting to put 117 in, we have a circumvention for the public certs , as we can just supply them.

**Comment by grzegorz8** [comment]
> @kristoffSC Friendly nudge. I assume this one is going to be processed before #117. We are wanting to put 117 in, we have a circumvention for the public certs , as we can just supply them.

Krzysztof seems to be less available recently. @davidradl Could you take over the PR and apply your suggestions? Or rather  create a new PR. I'll be happy to review it.

**Comment by davidradl** [comment]
> > @kristoffSC Friendly nudge. I assume this one is going to be processed before #117. We are wanting to put 117 in, we have a circumvention for the public certs , as we can just supply them.
> 
> Krzysztof seems to be less available recently. @davidradl Could you take over the PR and apply your suggestions? Or rather create a new PR. I'll be happy to review it.

That sounds good. I will take this one over - unless I hear otherwise form @kristoffSC

**Comment by davidradl** [comment]
@grzegorz8 @kristoffSC looking into this.

**Comment by davidradl** [comment]
@kristoffSC  @grzegorz8 I have done quite a lot of testing and asking round our team what our requirements are. It looks like the best option for us is actually @kristoffSC s fix . I need to slightly amend it as it does not quite work as expected - due to the way serverTrustedCerts is handled- it defaults to "" which end up as an array of 1  "" element, so the default is not obtained when expected. I will amend the fix and put up for review.

**Comment by davidradl** [comment]
I have coded this in [https://github.com/getindata/flink-http-connector/pull/128](https://github.com/getindata/flink-http-connector/pull/128).

sorry I did not manage to keep you original commit @kristoffSC.

**Comment by grzegorz8** [comment]
The issue was resolved in https://github.com/getindata/flink-http-connector/pull/128. Extra tests were added in https://github.com/getindata/flink-http-connector/pull/130.
